streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -581,8 +580,8 @@ public void punctuate(final ProcessorNode node, final long timestamp, final Punc
         if (processorContext.currentNode() != null) {
             throw new IllegalStateException(format("%sCurrent node is not null", logPrefix));
         }
-
-        updateProcessorContext(new StampedRecord(DUMMY_RECORD, timestamp), node);
+        
+        updateProcessorContext(new StampedRecord(new ConsumerRecord<>(ProcessorContextImpl.NONEXIST_TOPIC, -1, -1L, null, null), timestamp), node);
```

Comment:
This is fine, but note that the new consumer record here is just a roundabout way to create a ProcessorRecordContext. It'd probably be better to just directly instantiate the context we want.

Reply:
From my understanding, neither the `ConsumerRecord` nor the `ProcessroRecordContext` are the issue, but the shared `Header` object -- it's just a "side effect" that creating a new `ConsumerRecord` creates an new `Header` object internally.

Reply:
Yes, that's more or less what I was concerned about. We are trying to achieve a specific result here (and brand-new 'Header' instance each time), but it's only achieved via a side effect.

On the other hand, I just took a closer look at the code, and I see that there's a better reason to keep using `ConsumerRecord`, namely that `updateProcessorContext` is used both for regular processing (with "real" records) and for punctuation here with this dummy record. It seems like a good idea to prevent those code paths from diverging, so I'm +1 with keeping this change as-is.